openfasoc/generators/curref-gen/generators/custom_functions.py

```diff
@@ -0,0 +1,821 @@
+    idplace << c_route(pdk, idplace.ports['Via_center_'+str(vias[3])+'_B_src_htop_met_E'], idplace.ports['SB_E'], viaoffset=(False,False))
+
+    if with_tie:
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS2[0]+'_gate_W'], idplace.ports[nombres_puertos_tie_s[pos_tie_center_s-2]])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS2[0]+'_source_W'], idplace.ports[nombres_puertos_tie_s[pos_tie_center_s-2]])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS2[0]+'_drain_W'], idplace.ports[nombres_puertos_tie_s[pos_tie_center_s-2]])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS1[0]+'_gate_W'], idplace.ports[nombres_puertos_tie_n[pos_tie_center_s+1]])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS1[0]+'_source_W'], idplace.ports[nombres_puertos_tie_n[pos_tie_center_s+1]])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS1[0]+'_drain_W'], idplace.ports[nombres_puertos_tie_n[pos_tie_center_s+1]])
+      if with_substrate_tap and deviceA_and_B=='pfet':
+        idplace << c_route(pdk, idplace.ports[nombres_puertos_tie_s_W[pos_tie_center_s]], idplace.ports['VDD2_VSS2_W'], cglayer='met3', extension=0.4)
+        idplace << c_route(pdk, idplace.ports[nombres_puertos_tie_n_W[pos_tie_center_s]], idplace.ports['VDD1_VSS1_W'], cglayer='met3', extension=0.4, viaoffset=(False,False))
+        idplace << L_route(pdk, idplace.ports[nombres_puertos_tap_s[-1]], idplace.ports['substratetap1_E'], vglayer="met1")
+        idplace << L_route(pdk, idplace.ports[nombres_puertos_tap_n[0]], idplace.ports['substratetap2_E'], vglayer="met1")
+      else:
+        idplace << L_route(pdk, idplace.ports[nombres_puertos_tie_n[pos_tie_center_s+1]], idplace.ports['VDD1_VSS1_W'])
+        idplace << L_route(pdk, idplace.ports[nombres_puertos_tie_s[pos_tie_center_s-2]], idplace.ports['VDD2_VSS2_W'])
+    else:
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS1[0]+'_gate_E'], idplace.ports['VDD1_VSS1_N'])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS1[0]+'_source_E'], idplace.ports['VDD1_VSS1_N'])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS1[0]+'_drain_E'], idplace.ports['VDD1_VSS1_N'])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS2[0]+'_gate_E'], idplace.ports['VDD2_VSS2_S'])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS2[0]+'_source_E'], idplace.ports['VDD2_VSS2_S'])
+      idplace << L_route(pdk, idplace.ports['D'+puerto_VDDSS2[0]+'_drain_E'], idplace.ports['VDD2_VSS2_S'])
+      if with_substrate_tap and deviceA_and_B=='pfet':
+        idplace << L_route(pdk, idplace.ports[nombres_puertos_tap_s[-1]], idplace.ports['substratetap1_W'], vglayer="met1")
+        idplace << L_route(pdk, idplace.ports[nombres_puertos_tap_n[0]], idplace.ports['substratetap2_W'], vglayer="met1")
+
+    component = rename_ports_by_orientation(idplace).flatten()
+    return component
+
+
+def create_tapring_onchip(
```

Comment:
Can someone of these components or functions be converted into separate elementary blocks or primitives? They could be used in other generators.